### ISSUE TITLE
CosineAnnealingLrScheduler displays non-standard behavior

### ISSUE THREAD

**Issue by fuzing** [tracel-ai/burn#5128]
I'm using the following setup:

```
    let cosine_config = CosineAnnealingLrSchedulerConfig::new(max_lr, 250)
        .with_min_lr(min_lr);
```
where ``max_lr = 1e-3`` and ``min_lr = 1e-6``

And get the following recurring half-cosine LR waveform:

<img width="1419" height="905" alt="Image" src="https://github.com/user-attachments/assets/8c49abf7-2acd-4403-a4f2-b09a6f222fd0" />

For pytorch, documentation suggests that once the cosine wave reaches its minima, the LR remains there for the duration (i.e. without restarting the cycle - although I'm unsure what happens in the case where the minima approaches 0.0 which would be an unfortunate place to continue training, as it would be for Burn without configuring a .with_min_lr(), as the default minima would be 0.0 also).

Burn's current implementation which cycles a half cosine wave between max and min does not seem like an acceptable or expected outcome, and clearly could contribute to training instability.

**Comment by fuzing** [comment]
For pytorch, a separate [CosineAnnealingWarmRestarts](https://docs.pytorch.org/docs/2.12/generated/torch.optim.lr_scheduler.CosineAnnealingWarmRestarts.html) strategy incorporates ``warm restarts`` that appear to act similarly to burn's current version of ``CosineAnnealingLR``. It does, however, include configuration of a restart multipler level which acts on the denominator of the cosine function (the ``T_mult`` parameter), which can act to vary restart cosine period if a value other than the default of ``1`` is used.

Despite slight ambiguities in pytorch's documentation for [CosineAnnealingLR](https://docs.pytorch.org/docs/2.12/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html), it would appear that upon reaching minima, there is no restart, according to this [note](https://github.com/pytorch/pytorch/blob/0d62256a2b23365f8e1604297eb23a6545102aa8/torch/optim/lr_scheduler.py#L1361) as one might expect.

**Comment by nathanielsimard** [comment]
Happy to review a PR with the correction to be more aligned with PyTorch.

**Comment by DeathSurfing** [comment]
Hi! First-time contributor here, so I just wanted to make sure I understand the issue correctly.

From what I understand, Burn's `CosineAnnealingLrScheduler` currently behaves like PyTorch's `CosineAnnealingWarmRestarts`, whereas users would expect the behavior of `CosineAnnealingLR` (i.e. decay to `min_lr` and then remain there without restarting).

So would the expected fix be to change the behavior of the existing `CosineAnnealingLrScheduler` to match PyTorch, rather than introducing a separate warm-restart scheduler or making restarts configurable?

**Comment by fuzing** [comment]
Hi @DeathSurfing -

Thanks for your interest in this.  Yes - I think you're correct.  Technically, the version with warm restarts should behave like burn's current one (which should probably be renamed to its pytorch CosineAnnealingWarmRestarts namesake), with the exception that there is a multiplier in the pytorch version that controls the frequency (looking at the pytorch code).   As I understand it, there's a little bit of ambiguity in the pytorch CosineAnnealingLrScheduler, in that the documentation indicates that once a half cycle of the cosine is complete, it should remain at the terminal value.  Looking at the python code, however, it seems that the cosine would continue through the full cycle (i.e. without an actual reset), and then just keep repeating.

Hopefully that makes sense.

To confuse things though, there appears to be no value in stalling out the cosine at zero, as this will effectively halt further progress in reaching a lower minima (and just wastes resources on subsequent epochs/iterations).   The pytorch version does have a "minimum learning rate" which I'm presuming is aimed at countering this issue.

**Comment by DeathSurfing** [comment]
Thanks for the clarification! One thing I'm still unsure about before proceeding with the implementation:

From looking at PyTorch's implementation, `CosineAnnealingLR` doesn't actually appear to clamp the learning rate to `min_lr` after `T_max`. If `step()` continues to be called, the cosine continues through its cycle and the learning rate starts increasing again, rather than staying at `min_lr`.

Given that, should Burn aim to match PyTorch's actual implementation (for compatibility), or should it instead clamp the learning rate to `min_lr` after `num_iters`

**Comment by fuzing** [comment]
I did notice the inconsistencies in the pytorch docs vs. the Python code, but I think the intent for CosineAnnealingLR is to hold steady at ``min_lr`` once it is reached after ``num_iters``.   The question is what to do when the default ``min_lr`` is used (i.e. 0.0).   A learning rate of zero clearly halts further convergence.  My gut tells me that the default ``min_lr`` of 0.0 should still be adopted regardless, but that the documentation should highlight this danger (a consumer can override the default ``min_lr`` to prevent this situation after all).   The reason that I think simply continuing with the cosine beyond PI/2, increasing the L/R, is not such a great idea, is that actually increasing the learning rate is most often NOT what you'd want/expect.

For most use-cases ``num_iters`` needs to be precomputed anyway, such that ``min_lr`` is reached right at the end of a specific number of iterations/epochs.   Reinforcing my point from above, if one were to extend the training, I think it would be preferable to continue at ``min_lr`` instead of continuing with the cosine wave beyond PI/2 (which will increase the L/R).

For CosineAnnealingWarmRestarts, it is intended that the cosine jerks back to the initial value at cos(0) (albeit at a different frequency), so that false minima plateaus can be (hopefully) mitigated.  Personally, I find this one a little strange, as I'd imagine having the ability to vary both the frequency AND the amplitude would be most valuable to achieve this.

On the philosophical question of whether burn should aim to mimick pytorch/TF, that would be a question for one of the core team members.  In my opinion, burn would mimick pytorch/TF to minimize adoption hassles and match standard expectations, but would break with tradition where there are clear performance or other wins to be had.

Lastly, I'll be happy to contribute directly with PRs once my rust skill are a little further developed (I'm currently only about 6 weeks into the journey)
